Extend troubleshooting doc (#2072)
nginxinc/nginx-gateway-fabric: site/content/how-to/monitoring/troubleshooting.md

This topic describes possible issues users might encounter when using NGINX Gateway Fabric. When possible, suggested workarounds are provided.

> ### General Troubleshooting

Reviewer: ```suggestion
### General troubleshooting
```

Sentence case.

> When attempting to diagnose a problem or get support, there are a few important data points that can be collected to help with understanding what issues may exist.

Reviewer: ```suggestion
When investigating a problem or requesting help, there are important data points that can be collected to help understand what issues may exist.
```
Shortening the sentence. Support has specific implications so I favour avoiding the word unless it's in relation to an official support channel.

> ##### Resource Status

Reviewer: ```suggestion
##### Resource status
```

Sentence case.

> To get the status of a resource, use `kubectl describe`. For example, to check the status of the `coffee` HTTPRoute, which has an error:

Reviewer: ```suggestion
To check the status of a resource, use `kubectl describe`. This example checks the status of the `coffee` HTTPRoute, which has an error:
```

```shell
kubectl describe httproutes.gateway.networking.k8s.io coffee [-n namespace]
```

```text
...
Status:
  Parents:
    Conditions:
      Last Transition Time:  2024-05-31T17:20:51Z
      Message:               The route is accepted
      Observed Generation:   4
      Reason:                Accepted
      Status:                True
      Type:                  Accepted
      Last Transition Time:  2024-05-31T17:20:51Z
      Message:               spec.rules[0].backendRefs[0].name: Not found: "bad-backend"
      Observed Generation:   4
      Reason:                BackendNotFound
      Status:                False
      Type:                  ResolvedRefs
    Controller Name:         gateway.nginx.org/nginx-gateway-controller
    Parent Ref:
      Group:         gateway.networking.k8s.io
      Kind:          Gateway
      Name:          gateway
      Namespace:     default
      Section Name:  http
```

> If a resource has any errors relating to its configuration or relation to other resources, it is likely that those errors will be contained within the status. The `ObservedGeneration` in the status should match the `ObservedGeneration` of the resource. Otherwise, this could mean that the resource wasn't processed yet or the status failed to update.

Reviewer: ```suggestion
If a resource has errors relating to its configuration or relationship to other resources, they can likely be read in the status. The `ObservedGeneration` in the status should match the `ObservedGeneration` of the resource. Otherwise, this could mean that the resource hasn't been processed yet or that the status failed to update.
```
This is mostly tense consistency.

##### Events

> Events may be created by NGINX Gateway Fabric or other Kubernetes components that could indicate system or configuration issues. To see events:

Reviewer: ```suggestion
Events created by NGINX Gateway Fabric or other Kubernetes components could indicate system or configuration issues. To see events:
```
Some unnecessary words.

```shell
kubectl get events [-n namespace]
```

For example, a warning event when the NginxGateway configuration CRD is deleted:

```text
kubectl -n nginx-gateway get event
LAST SEEN   TYPE      REASON              OBJECT                                           MESSAGE
5s          Warning   ResourceDeleted     nginxgateway/ngf-config                          NginxGateway configuration was deleted; using defaults
```

##### Logs

> Logs of the NGINX Gateway Fabric control plane and data plane can contain information that isn't otherwise reported in status or events. These could include errors in processing or passing traffic.

Reviewer: ```suggestion
Logs from the NGINX Gateway Fabric control plane and data plane can contain information that isn't available to status or events. These can include errors in processing or passing traffic.
```

To see logs for the control plane container:

```shell
kubectl -n nginx-gateway logs <ngf-pod-name> -c nginx-gateway
```

To see logs for the data plane container:

```shell
kubectl -n nginx-gateway logs <ngf-pod-name> -c nginx
```

> You can also see the logs of a container that has crashed or been killed, by specifying the `-p` flag with the above commands.

Reviewer: ```suggestion
You can see logs for a crashed or killed container by adding the `-p` flag to the above commands.
```

This topic describes possible issues users might encounter when using NGINX Gateway Fabric. When possible, suggested workarounds are provided.

> ### **General Troubleshooting**

Reviewer 2: do we need to apply this technique of placing header sections in `**` across all docs? 
It's better to be consistent

Author: No, I just thought that the headers in this particular doc might run together too much, so I wanted to create some distinction. I'll actually revert this and use a different header type for the nested ones.

When attempting to diagnose a problem or get support, there are a few important data points that can be collected to help with understanding what issues may exist.

#### Resource Status

To get the status of a resource, use `kubectl describe`. For example, to check the status of the `coffee` HTTPRoute:

```shell
kubectl describe httproutes.gateway.networking.k8s.io coffee [-n namespace]
```

```text
...

> Status:

Reviewer 2: would it make sense to show an example of an error rather than a successful case?

  Parents:
    Conditions:
      Last Transition Time:  2024-05-31T16:22:26Z
      Message:               The route is accepted
      Observed Generation:   1
      Reason:                Accepted
      Status:                True
      Type:                  Accepted
      Last Transition Time:  2024-05-31T16:22:26Z
      Message:               All references are resolved
      Observed Generation:   1
      Reason:                ResolvedRefs
      Status:                True
      Type:                  ResolvedRefs
    Controller Name:         gateway.nginx.org/nginx-gateway-controller
    Parent Ref:
      Group:         gateway.networking.k8s.io
      Kind:          Gateway
      Name:          gateway
      Namespace:     default
      Section Name:  http
```

> If a resource has any errors relating to its configuration or relation to other resources, it is likely that those errors will be contained within the status.

Reviewer 2: I wonder if we should also mention that observed generation in conditions should match observed generation of the resource. if it doesn't match, this means NGF either didn't process the resource update yet or it failed to update its status.

#### Events

Events may be created by NGINX Gateway Fabric or other Kubernetes components that could indicate system or configuration issues. To see events:

```shell

> kubectl get events [-n namespace]

Reviewer 2: perhaps we can have an example from NGF?

```

#### Logs

Logs of the NGINX Gateway Fabric control plane and data plane can contain information that isn't otherwise reported in status or events. These could include errors in processing or passing traffic.

> To see logs for the control plane container (replacing the name of the deployment if necessary):

Reviewer 2: since users can scale pods, how would it work when there > 1 pod?

Author: Good point, I'll use pod instead of deployment
